lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -914,6 +936,106 @@ makeChange minCoinValueFor requiredCost mExtraCoinSource inputBundles outputBund
             ((`Set.notMember` knownAssetIds) . fst)
             (TokenMap.toFlatList tokens)
 
+-- | Assigns coin quantities to a list of pre-computed asset change maps.
+--
+-- Each pre-computed asset change map must be paired with the original coin
+-- value of its corresponding output.
+--
+-- This function:
+--
+--    - expects the list of pre-computed asset change maps to be sorted in an
+--      order that ensures all empty token maps are at the start of the list.
+--
+--    - attempts to assign a minimum ada quantity to every change map, but
+--      iteratively drops empty change maps from the start of the list if the
+--      amount of ada is insufficient to cover them all.
+--
+--    - continues dropping empty change maps from the start of the list until
+--      it is possible to assign a minimum ada value to all remaining entries.
+--
+--    - returns a list that is identical in length to the input list if (and
+--      only if) it was possible to assign a minimum ada quantity to all change
+--      maps.
+--
+--    - returns a list that is shorter than the input list if it was only
+--      possible to assign a minimum ada quantity to a suffix of the given
+--      list.
+--
+--    - returns 'Nothing' if (and only if) there was not enough ada available
+--      to assign the minimum ada quantity to all non-empty change maps.
+--
+assignCoinsToChangeMaps
+    :: Coin
+    -- ^ The total quantity of ada available, including any extra source of ada.
+    -> (TokenMap -> Coin)
+    -- ^ A function to calculate the minimum required ada quantity for any
+    -- token map.
+    -> NonEmpty (TokenMap, Coin)
+    -- ^ A list of pre-computed asset change maps paired with original output
+    -- coins, sorted into an order that ensures all empty token maps are at the
+    -- start of the list.
+    -> Maybe [TokenBundle]
+    -- ^ Resulting change bundles, or 'Nothing' if there was not enough ada
+    -- available to assign a minimum ada quantity to all non-empty token maps.
+assignCoinsToChangeMaps adaAvailable minCoinFor pairsAtStart =
+    loop adaRequiredAtStart pairsAtStart
+  where
+    loop !adaRequired !pairsNonEmpty = case pairsNonEmpty of
```

Comment:
Isn't it a bit confusing that `adaRequired` comes as an argument here? I'd expect whatever branch needs it to re-calculate it in its context. That's the case for the first branch only.

Reply:
> Isn't it a bit confusing that adaRequired comes as an argument here?

We could re-calculate it on every iteration, but presumably that would make it necessary to compute the minimum values across all remaining token maps in the list, which would lead to quadratic run-time in the length of the list, unless you're thinking of something else?

Here I've opted to compute the total requirement up front, and then just reduce it on each iteration by the correct amount (which should be equal to the minimum ada value of the map that we drop).